File: leetcode/Easy/0067-add-binary/0067-add-binary.cpp
/*
 Problem: 67. Add Binary
 Language: cpp
 Runtime: 0 ms (100.00%)
 Memory: 8.9 MB (82.65%)
 Tags: Math, String, Bit Manipulation, Simulation
*/
class Solution {
public:
    string addBinary(string a, string b) {
        int pa = a.size() - 1;
        int pb = b.size() - 1;
        string ans = "";
        int carry = 0;

        while (pa >= 0 || pb >= 0 || carry) {
            int sum = carry;
            if (pa >= 0)
                sum += a[pa--] - '0';
            if (pb >= 0)
                sum += b[pb--] - '0';
            ans += (sum % 2) + '0';
            carry = sum / 2;
        }
        
        reverse(ans.begin(), ans.end());
        
        return ans;
    }
};

File: leetcode/Easy/0067-add-binary/0067-add-binary.rs
/*
 Problem: 67. Add Binary
 Language: rust
 Runtime: 0 ms (100.00%)
 Memory: 2.2 MB (16.32%)
 Tags: Math, String, Bit Manipulation, Simulation
*/
impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        let bytes_a = a.as_bytes();
        let bytes_b = b.as_bytes();

        let mut i: i32 = bytes_a.len() as i32 - 1;
        let mut j: i32 = bytes_b.len() as i32 - 1;

        let mut carry = 0;
        let mut result: Vec<char> = Vec::new();

        while i >= 0 || j >= 0 || carry > 0 {
            let mut sum = carry;

            if i >= 0 {
                sum += (bytes_a[i as usize] - b'0') as i32;
                i -= 1;
            }

            if j >= 0 {
                sum += (bytes_b[j as usize] - b'0') as i32;
                j -= 1;
            }

            result.push(char::from(b'0' + (sum % 2) as u8));
            carry = sum / 2;
        }

        result.iter().rev().collect()
    }
}
